1183 am file new

ryangrundy7 opened the pull request:
# Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
To be able to add addresses confirmed by AR as new cases, a new bulk processor needs to be created to create all the new cases we're going to be getting. The new cases should send a message to Field and RH and it doesn't create any uac-qids for the new cases.

# What has changed
<!--- What manifest changes have been made? -->
- Added new bulk processor for creating new addresses from AR
- Added new validators for new addresses schema

# How to test?
<!--- Describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->
<!--- Are there any automated tests that mean changes don't need to be manually tested? -->
- To test, run with other branches on pr (case-processor, terraform and kubernetes). Once the new bucket is created, you should be able to upload a file in the bucket and run the new bulk processor. This should create new cases and send them to field and RH. Also check to see to make sure no uac-qids are created from them.

# Links
<!--- Add any links to issues (Jira, Trello, GitHub Issues) -->
<!--- Links to any documentation. -->
<!--- Links to any related PRs. -->
[Trello](https://trello.com/c/ATLrBzMk)
[Kubernetes](https://github.com/ONSdigital/census-rm-kubernetes/pull/288)
[case-processor](https://github.com/ONSdigital/census-rm-case-processor/pull/181)
[Terraform](https://github.com/ONSdigital/census-rm-terraform/pull/186)

ryangrundy7 committed:
Added new address bulk processor

ryangrundy7 committed:
Pass in the column into the kwargs for the validators
add tests for new_address_processor and new validators
updated treatment codes for processor

ryangrundy7 committed:
Added alias for bulk new address

ryangrundy7 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [AM - File - New (13)](https://trello.com/c/ATLrBzMk/1183-am-file-new-13)

AdamHawtin commented on bulk_processing/validators.py:
```diff
@@ -48,3 +49,88 @@ def validate(case_id, **kwargs):
+    return validate
+
+
+def unique():
+    previous_values = set()
+
+    def validate(value, **_kwargs):
+        if value in previous_values:
+            raise Invalid(f'Value "{value}" is not unique')
+        previous_values.add(value)
+
+    return validate
+
```
I don't think this is used at all, can be removed

ryangrundy7 committed:
Updated readme

AdamHawtin commented on test/bulk_processing/test_new_address_processor.py:
```diff
@@ -0,0 +1,66 @@
+    events = new_address_processor.build_event_messages(new_address_row)
+
+    # Then
+    assert len(events) == 1, 'Should build one and only 1 refusal event message'
```
Description says refusal event

AdamHawtin commented on test/bulk_processing/test_new_address_processor.py:
```diff
@@ -0,0 +1,66 @@
+    assert events[0]['collectionExerciseId'] == Config.COLLECTION_EXERCISE_ID
+
+
+@patch('bulk_processing.bulk_processor.storage')
```
Could use fixture for this instead, it's repeated a lot

AdamHawtin commented on test/bulk_processing/test_new_address_processor.py:
```diff
@@ -0,0 +1,66 @@
+def test_new_address_validation_errors_doesnt_match(_patched_storage_client):
+    result = BulkProcessor(NewAddressProcessor()).find_header_validation_errors(file_headers_errors)
+
+    assert 'ARID' in result.description
```
Is 'ARID' an unexpected header here? Could do with a comment on why it's supposed to be in the error description to avoid confusion since it was previously a valid, mandatory column.

AdamHawtin commented on bulk_processing/new_address_processor.py:
```diff
@@ -0,0 +1,94 @@
+        'MSOA': [mandatory(), max_length(9), no_padding_whitespace()],
+        'LAD': [mandatory(), max_length(9), no_padding_whitespace()],
+        'REGION': [mandatory(), max_length(9), no_padding_whitespace(), region_matches_treatment_code()],
+        'HTC_WILLINGNESS': [mandatory(), in_set({'0', '1', '2', '3', '4', '5'}, label='HTC_WILLINGNESS'),
```
We should decide on one convention for the labels since they get passed into the error details we feed back. Using the column header makes sense to me.

ryangrundy7 committed:
Removed unused unique validator
updated test comments
removed TODO from config

NickGrantX commented on bulk_processing/new_address_processor.py:
```diff
@@ -0,0 +1,94 @@
+        'ESTAB_UPRN': [mandatory(), max_length(12), numeric(), mandatory(), no_padding_whitespace()],
+        'ADDRESS_TYPE': [mandatory(), in_set({'HH', 'CE', 'SPG'}, label='address_types'), no_padding_whitespace()],
+        'ESTAB_TYPE': [mandatory(), in_set(ESTAB_TYPES, label='estab types'), max_length(255), no_padding_whitespace()],
+        'ADDRESS_LEVEL': [mandatory(), in_set({'E', 'U'}, label='address levels'), no_padding_whitespace()],
```
we should probably boost the validation now that we're opening the floodgates to chaos monkeys - we can't have a 'U' level on an 'HH' address type, for example

ryangrundy7 replied:
I'll create a card to look at the extra validation we could put onto this and in the sample validator. This is card was only meant for copying those over

NickGrantX commented on bulk_processing/new_address_processor.py:
```diff
@@ -0,0 +1,94 @@
+                            no_padding_whitespace()],
+        'HTC_DIGITAL': [mandatory(), in_set({'0', '1', '2', '3', '4', '5'}, label='htc digital'),
+                        no_padding_whitespace()],
+        'TREATMENT_CODE': [mandatory(), in_set(TREATMENT_CODES, label='treatment codes'), no_padding_whitespace()],
```
we should probably validate that the treatment code is valid for the case type etc

ryangrundy7 replied:
I'll create a card to look at the extra validation we could put onto this and in the sample validator. This is card was only meant for copying those over.

NickGrantX commented on bulk_processing/new_address_processor.py:
```diff
@@ -0,0 +1,94 @@
+        'FIELDOFFICER_ID': [max_length(13), no_padding_whitespace()],
+        'CE_EXPECTED_CAPACITY': [numeric(), max_length(4), no_padding_whitespace(), ce_u_has_expected_capacity()],
+        'CE_SECURE': [mandatory(), in_set({'0', '1'}, label='CE Secure'), no_padding_whitespace()],
+        'PRINT_BATCH': [numeric(), max_length(2), no_padding_whitespace()]
```
shouldn't be `0` or `00`

ryangrundy7 replied:
I'll create a card to look at the extra validation we could put onto this and in the sample validator. This is card was only meant for copying those over

NickGrantX reviewed (changes requested):
See comments - validation is more hardcore, but it needs to be, because we could be sent any old garbage via the bulk process

AdamHawtin commented:
@NickGrantONS We said we would keep the validation here the same as the sample loader. I don't disagree with any of your suggestions, but I think we should keep the two in sync and add anything extra here to the sample loader validation tool too.

ryangrundy7 committed:
Labels are now column names
fixed bug when header is missing, it created an error file consisting of just header and deleted contents

ryangrundy7 committed:
Added column to error messages to make it easier to read

AdamHawtin reviewed (approved):
LGTM

ryangrundy7 committed:
Updated processor to fail if action_plan_id and collection_exercise_id aren't set

liamtoozer commented on bulk_processing/new_address_processor.py:
```diff
@@ -0,0 +1,98 @@
+
+    file_prefix = Config.BULK_NEW_ADDRESS_FILE_PREFIX
+    routing_key = Config.NEW_ADDRESS_EVENT_ROUTING_KEY
+    exchange = ''
```
Think it needs to grab the exchange from the config here?

ryangrundy7 replied:
I looked at sample loader and it doesn't use the exchange and just puts it on the queue directly with the routing key

liamtoozer replied:
Ah okay cool, that's good to know 👍

NickGrantX reviewed (approved):
Agree that validation can be beefed up on another ticket
